autoload/clang_format.vim

```diff
@@ -291,6 +307,95 @@ endfunction
 function! clang_format#disable_auto_format() abort
     let g:clang_format#auto_format = 0
 endfunction
+" s:chdir will change the directory respecting
+" local/tab-local/global directory settings.
+function! s:chdir(dir)
+    " This is a dirty hack to fix tcd breakages on neovim.
+    " Future work should be based on nvim API.
+    if exists(':tcd')
+        let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+    else
+        let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
+    endif
+    execute chdir fnameescape(a:dir)
+endfunction
+
+" s:strip: helper function to strip a string
+function! s:strip(string)
+    return substitute(a:string, '^\s*\(.\{-}\)\s*\n\=$', '\1', '')
+endfunction
+
+" this function will try to format only buffer lines diffing from git index
+" content.
+" If the file is untracked (not in a git repo or not tracked in a git repo),
+" it returns 1.
+" If the format succeeds, it returns 0.
+function! clang_format#do_auto_format_git_diff()
+    let dir = getcwd()
+    let cur_file = expand("%:p")
+    let cur_file_path = isdirectory(cur_file) ? cur_file : fnamemodify(cur_file, ":h")
+    try
+        call s:chdir(cur_file_path)
+        let top_dir=s:strip(system(
+                    \ g:clang_format#git." rev-parse --show-toplevel"))
+        if v:shell_error != 0
+            throw "untracked"
```

Comment:
Please use more informational error message. It's better for maintenance even if the exception is not thrown to outside.
```suggestion
            throw "clang-format: untracked file by git"
```